benchmarks/nemo/tracer_advection/scripts/acc_mixed_explicit_data_movement_trans.py

```diff
@@ -0,0 +1,207 @@
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
+
+'''A transformation script that seeks to apply OpenACC KERNELS
+directives to NEMO style code.  In order to use
+it you must first install PSyclone. See README.md in the top-level
+psyclone directory.
+
+Once you have psyclone installed, this may be used by doing:
+
+ $ psyclone -api nemo -s ./kernels_trans.py some_source_file.f90
+
+This should produce a lot of output, ending with generated
+Fortran. Note that the Fortran source files provided to PSyclone must
+have already been preprocessed (if required).
+
+The transformation script attempts to insert Kernels directives at the
+highest possible location(s) in the schedule tree (i.e. to enclose as
+much code as possible in each Kernels region). However, due to
+limitations in the PGI compiler, we must take care to exclude certain
+nodes (such as If blocks) from within Kernel regions. If a proposed
+region is found to contain such a node (by the ``valid_kernel``
+routine) then the script moves a level down the tree and then repeats
+the process of attempting to create the largest possible Kernel
+region.
+
+'''
+
+from __future__ import print_function
+from psyclone.domain.nemo.transformations import NemoAllArrayRange2LoopTrans
+from psyclone.errors import InternalError
+from psyclone.psyir.nodes import Assignment, CodeBlock, Call, Literal, Loop, \
+    ACCLoopDirective
+from psyclone.psyir.transformations import ACCUpdateTrans
+from psyclone.transformations import ACCLoopTrans, TransformationError, \
+    ACCKernelsTrans, ACCEnterDataTrans
+
+
+COLLAPSE_LOOPS = False
+EXPAND_IMPLICIT_LOOPS = False
+
+# Get the PSyclone transformations we will use
+ARRAY_RANGE_TRANS = NemoAllArrayRange2LoopTrans()
+
+
+def trans(psy):
+    '''A PSyclone-script compliant transformation function. Applies
+    OpenACC 'kernels' and 'data' directives to NEMO code.
+
+    :param psy: The PSy layer object to apply transformations to.
+    :type psy: :py:class:`psyclone.psyGen.PSy`
+    '''
+
+    print("Invokes found:\n{0}\n".format(
+        "\n".join([str(name) for name in psy.invokes.names])))
+
+    for invoke in psy.invokes.invoke_list:
+
+        sched = invoke.schedule
+        if not sched:
+            print("Invoke {0} has no Schedule! Skipping...".
+                  format(invoke.name))
+            continue
+        sched.view()
+
+        if EXPAND_IMPLICIT_LOOPS:
```

Comment:
I'm a bit confused here. I thought 'mixed' meant a combination of the loop transformations with the use of ACC kernels but I can't see that this script does any loop transformations? (`COLLAPSE_LOOPS` and `EXPAND_IMPLICIT_LOOPS` are both `False`.)

Reply:
Yeah sorry, this was a remain of code before I decided that for benchmarking it would be easier to use calling different scripts instead of parameters. I removed all this now and always use the utils functions.